moc_openstack_api_reporting.py

```diff
@@ -504,51 +546,168 @@ def process_compute_events(openstack_conn, script_datetime, openstack_data, serv
+        "user_id": volume_data.user_id,
+        "user_name": openstack_data["user_dict"][volume_data.user_id]["name"],
+        "domain": project_rec["domain_id"],
+    }
+
+    event_to_status = {
+        "volume.create.start": "creating",
+        "volume.create.end": "available",
+        "volume.attach.start": "",
+        "volume.attach.end": "",
+        "volume.delete.start": "",
+        "volume.delete.end": "",
+    }
+
+    if volume_event_template["event_type"].startswith("volume.create"):
+        volume_event_template["generated"] = volume_data.created_at
+
+    event_list = convert_to_ceilometer_event_types(volume_event_template)
+    for ceilometer_event in event_list:
+        ceilometer_event["status"] = event_to_status[ceilometer_event["event_type"]]
+
+    return event_list
+
+
+def process_volume_events(openstack_data, cluster_state):
+    """collects and processes event data"""
+    events_by_date = {}
+    # so that we can tell if the vol not present
+    for vol_id in cluster_state["vol_timestamps"]:
+        cluster_state["vol_timestamps"][vol_id]["updated"] = 0
+
+    volume_dict = openstack_data["volume_dict"]
+    events_by_date = None
+    for vol_id, volume_data in volume_dict.items():
+        if vol_id not in cluster_state["vol_timestamps"]:
+            cluster_state["vol_timestamps"][vol_id] = {}
+            cluster_state["vol_timestamps"][vol_id]["updated"] = 0
+
+        # consider using volume_data.updated_at for resize events
+        # event_datetime = datetime.datetime.fromisoformat(volume_data.updated_at)
+        last_run_datetime = None
+        if cluster_state["last_run_timestamp"]:
+            last_run_datetime = datetime.datetime.fromisoformat(cluster_state["last_run_timestamp"])
+        vol_create_datetime = datetime.datetime.fromisoformat(volume_data.created_at)
+
+        if last_run_datetime < vol_create_datetime:
+            volume_event_list = create_volume_event(openstack_data, vol_id, "volume.create")
+            events_by_date = merge_cache_with_current_data(volume_event_list, events_by_date)
+            cluster_state["vol_timestamps"][vol_id]["updated"] = 1
+
+        if volume_data.status == "deleted":
+            del cluster_state["vol_timestamps"][vol_id]
+            volume_event_list = create_volume_event(openstack_data, vol_id, "volume.delete")
+            events_by_date = merge_cache_with_current_data(openstack_data, events_by_date)
+            cluster_state["vol_timestamps"][vol_id]["updated"] = 1
 
     return events_by_date
 
 
 def read_json_file(filename, default):
     """reads a json file and returns the data"""
-    try:
-        with open(filename, "r", encoding="utf-8") as file:
-            return json.load(file)
-    except IOError:
-        pass
-    except ValueError:
-        # Ignore badly formatted file or empty file
-        pass
+    with open(filename, "r", encoding="utf-8") as file:
+        return json.load(file)
     return default
```

Comment:
This line will never be reached. The `default` value is never used in this function.